okhttp-sse/src/main/java/okhttp3/internal/sse/RealEventSource.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.sse;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.EventListener;
+import okhttp3.EventSource;
+import okhttp3.EventSourceListener;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.internal.Util;
+import okio.BufferedSource;
+
+public final class RealEventSource
+    implements EventSource, ServerSentEventReader.Callback, Callback {
+
+  private final Request request;
+  private final EventSourceListener listener;
+
+  private Call call;
+
+  public RealEventSource(Request request, EventSourceListener listener) {
+    this.request = request;
+    this.listener = listener;
+  }
+
+  public void connect(OkHttpClient client) {
+    client = client.newBuilder()
+        .eventListener(EventListener.NONE)
+        .build();
+    call = client.newCall(request);
+    call.enqueue(this);
+  }
+
+  @Override public void onResponse(Call call, Response response) {
+    //noinspection ConstantConditions main body is never null
+    BufferedSource source = response.body().source();
+    ServerSentEventReader reader = new ServerSentEventReader(source, this);
+
+    response = response.newBuilder()
+        .body(Util.EMPTY_RESPONSE)
+        .build();
+
+    try {
+      listener.onOpen(this, response);
+      while (reader.processNextEvent()) {
+      }
+    } catch (Exception e) {
```

Comment:
Exception here seem counter to the normal OkHttp opinionated IOException or GTFO approach :) . Do you want to swallow NPE etc here?